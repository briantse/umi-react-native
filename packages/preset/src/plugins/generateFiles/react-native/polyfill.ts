import { IApi } from '@umijs/types';

const CONTENT = `{{{ polyfills }}}
import 'react-native/Libraries/Core/InitializeCore.js';

// @ts-ignore
if (global.window === undefined) {
  // @ts-ignore
  global.window = global;
}

// disable hot module replacement
// @ts-ignore
if (module.hot) {
  // @ts-ignore
  module.hot.accept = function() {
  
  };
}

`;

export default (api: IApi) => {
  const {
    utils: { Mustache, semver },
  } = api;

  // api.addPolyfillImports(() => [
  //   {
  //     source: './react-native/polyfill',
  //   },
  // ]);

  // RN 不需要 @babel/polyfill，这里通过 alias 替换 @umijs/preset-built-in 中的 polyfill
  api.chainWebpack((memo) => {
    memo.resolve.alias.set('./core/polyfill', './react-native/polyfill');
    return memo;
  });

  api.onGenerateFiles(() => {
    const polyfills: string[] = [];
    if (semver.lt(api.config?.reactNative?.version, '0.60.0')) {
      polyfills.push(
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Object.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/console.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/error-guard.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Number.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/String.prototype.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Array.prototype.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Array.es6.js'),
        require.resolve('@haul-bundler/preset-0.59/vendor/polyfills/Object.es7.js'),
      );
    } else {
      polyfills.push(
        'react-native/Libraries/polyfills/console.js',
        'react-native/Libraries/polyfills/error-guard.js',
        'react-native/Libraries/polyfills/Object.es7.js',
      );
    }
    api.writeTmpFile({
      path: 'react-native/polyfill.ts',
      content: Mustache.render(CONTENT, {
        polyfills: polyfills.map((it: string) => `import '${it}';`).join('\n'),
      }),
    });
  });
};
