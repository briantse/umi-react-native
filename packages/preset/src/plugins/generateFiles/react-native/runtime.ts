import { IApi } from '@umijs/types';
import { join } from 'path';

const runtimeTpl = `import {AppRegistry} from 'react-native';

export function render(clientRender: () => any, args: {hot?: boolean} = {}) {
  AppRegistry.registerComponent('{{{ appKey }}}', () => clientRender);
}

`;

export default (api: IApi) => {
  api.onGenerateFiles(() => {
    // 用户配了 mountElementId 就不再启动 RN 应用
    if (api.config.mountElementId) return;
    api.addRuntimePlugin(() => [join(api.paths.absTmpPath!, 'react-native', 'runtime.ts')]);
    api.writeTmpFile({
      path: 'react-native/runtime.ts',
      content: api.utils.Mustache.render(runtimeTpl, {
        appKey: api.config?.reactNative?.appKey,
      }),
    });
  });
};
