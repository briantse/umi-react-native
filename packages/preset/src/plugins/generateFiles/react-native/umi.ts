/**
 * @file 防止加载umi包 Common JS格式的代码
 * @description umi for WEB 使用webpack treeShaking，运行时加载的是ES Module格式的代码：umi/dist/index.esm.js，没问题。
 * 但 RN 中 无论 haul 还是 metro，都没有treeShaking，会加载umi Common JS代码：umi/index.js，其中包含了大量 Node 工具类库。
 * 在 haul 构建时甚至会导致Out of memory。
 */
import { IApi } from '@umijs/types';

const CONTENT = `export * from 'umi-react-native-runtime';
export * from '../core/umiExports';

`;

export default (api: IApi) => {
  api.onGenerateFiles(() => {
    api.writeTmpFile({
      path: 'react-native/umi.ts',
      content: CONTENT,
    });
  });
};
