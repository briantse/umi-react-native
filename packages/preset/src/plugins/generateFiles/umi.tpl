{{{ polyfillImports }}}
{{{ importsAhead }}}
import { plugin } from './core/plugin';
import { ApplyPluginsType } from '{{{ runtimePath }}}';
import { renderClient } from '{{{ rendererPath }}}';
{{{ imports }}}

{{{ entryCodeAhead }}}

const getClientRender = (args: { hot?: boolean } = {}) => plugin.applyPlugins({
  key: 'render',
  type: ApplyPluginsType.compose,
  initialValue: () => {
    return renderClient({
      // @ts-ignore
      routes: require('./core/routes').routes,
      plugin,
    });
  },
  args,
});

const clientRender = getClientRender();
export default clientRender();

{{{ entryCode }}}

// hot module replacement
// @ts-ignore
if (module.hot) {
  // @ts-ignore
  module.hot.accept('./core/routes', () => {
    getClientRender({ hot: true })();
  });
}
