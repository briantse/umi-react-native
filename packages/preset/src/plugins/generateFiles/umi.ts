import { existsSync, readFileSync } from 'fs';
import { join, dirname } from 'path';
import { EOL } from 'os';
import { IApi } from '@umijs/types';
import { winPath } from '@umijs/utils';
import assert from 'umi-react-native-cli/lib/utils/assert';

export function importsToStr(imports: { source: string; specifier?: string }[]) {
  return imports.map((imp) => {
    const { source, specifier } = imp;
    if (specifier) {
      return `import ${specifier} from '${winPath(source)}';`;
    } else {
      return `import '${winPath(source)}';`;
    }
  });
}

export default function (api: IApi) {
  const {
    utils: { Mustache },
  } = api;

  api.onGenerateFiles(async (args) => {
    const umiTpl = readFileSync(join(__dirname, 'umi.tpl'), 'utf-8');
    const rendererPath = await api.applyPlugins({
      key: 'modifyRendererPath',
      type: api.ApplyPluginsType.modify,
      initialValue: require.resolve('@umijs/renderer-react'),
    });
    let appKey = api.config.appKey;
    if (!appKey) {
      const appJsonFilename = join(api.paths.cwd || process.cwd(), 'app.json');
      if (!existsSync(appJsonFilename)) {
        throw new Error(`${appJsonFilename} was not exists.`);
      }
      try {
        const appJson = JSON.parse(readFileSync(appJsonFilename, 'utf8'));
        appKey = appJson.name;
      } catch (ignored) {}
    }
    if (!appKey) {
      throw new TypeError(
        `"appKey" 未配置。${EOL}1. 请在UMI配置文件（比如：.umirc.js）中添加"appKey"字段，并设置一个值；${EOL}2. 或者在工程根目录下的 app.json 文件中设置"name"字段和值。${EOL}小贴士:${EOL}"appKey"即RN JS代码域中: "AppRegistry.registerComponent(appKey, componentProvider);"的第一个参数；也是iOS/Android原生代码中加载bundle时所需的"moduleName"，其值通常是使用 react-native 命令行工具初始化新建RN工程时指定的项目名称。`,
      );
    }
    api.writeTmpFile({
      path: 'umi.ts',
      content: Mustache.render(umiTpl, {
        appKey,
        rendererPath: winPath(rendererPath),
        runtimePath: winPath(dirname(require.resolve('@umijs/runtime/package.json'))),
        entryCode: (
          await api.applyPlugins({
            key: 'addEntryCode',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          })
        ).join('\r\n'),
        entryCodeAhead: (
          await api.applyPlugins({
            key: 'addEntryCodeAhead',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          })
        ).join('\r\n'),
        polyfillImports: importsToStr(
          await api.applyPlugins({
            key: 'addPolyfillImports',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          }),
        ).join('\r\n'),
        importsAhead: importsToStr(
          await api.applyPlugins({
            key: 'addEntryImportsAhead',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          }),
        ).join('\r\n'),
        imports: importsToStr(
          await api.applyPlugins({
            key: 'addEntryImports',
            type: api.ApplyPluginsType.add,
            initialValue: [],
          }),
        ).join('\r\n'),
      }),
    });
  });
}
