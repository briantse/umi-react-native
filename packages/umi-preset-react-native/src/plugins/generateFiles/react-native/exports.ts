import { IApi } from 'umi';

const TEMPLATE = `{{#dynamicImport}}
import Multibundle from 'umi-react-native-multibundle';

export async function dynamicImportBundle(bundleName: string) {
  await Multibundle.loadBundle(bundleName);
  return Multibundle.getBundleExport(bundleName);
}

{{/dynamicImport}}
export {
  BackButton,
  AndroidBackButton,
} from 'react-router-native';

`;

export default (api: IApi) => {
  api.onGenerateFiles(() => {
    api.writeTmpFile({
      path: 'react-native/exports.ts',
      content: api.utils.Mustache.render(TEMPLATE, {
        dynamicImport: Boolean(api.config.dynamicImport),
      }),
    });
  });
  api.addUmiExports(() => [
    {
      exportAll: true,
      source: '../react-native/exports',
    },
  ]);
};
