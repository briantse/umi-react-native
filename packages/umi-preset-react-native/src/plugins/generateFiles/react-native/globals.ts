import { IApi } from 'umi';

const TPL = `{{{ globals }}}

`;

export default (api: IApi) => {
  const {
    utils: { Mustache, lodash },
  } = api;
  api.addPolyfillImports(() => {
    if (lodash.isPlainObject(api.config.define) && !lodash.isEmpty(api.config.define)) {
      return [
        {
          source: './react-native/globals',
        },
      ];
    }
    return [];
  });
  api.onGenerateFiles(() => {
    if (lodash.isPlainObject(api.config.define) && !lodash.isEmpty(api.config.define)) {
      api.writeTmpFile({
        path: 'react-native/globals.ts',
        content: Mustache.render(TPL, {
          globals: lodash
            .keysIn(api.config.define)
            .map(
              (key) =>
                `// @ts-ignore\nglobal['${key}'] = ${JSON.stringify(api.config.define && api.config.define[key])};`,
            )
            .join('\n'),
        }),
      });
    }
  });
};
