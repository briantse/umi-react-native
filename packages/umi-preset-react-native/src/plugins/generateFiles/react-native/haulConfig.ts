import { IApi } from 'umi';
import { dirname } from 'path';
import { bundlesToJSON, createFormatter, transformRoutesToBundle } from '../../../utils';

const CONTENT = `import _ from 'lodash';
import { makeConfig, withPolyfills } from '{{{ haulPresetPath }}}';

const transform = ({ config }) => {
  return _.defaultsDeep({{{ webpackConfig }}}, config);
};

export default makeConfig({
  templates: {
    filename: {
      ios: '[bundleName].ios.bundle',
    },
  },
  features: {
    multiBundle: 2,
  },
  bundles: {
    index: {
      entry: withPolyfills(
        ['@@/core/plugin', ...{{{ dependencies }}}],
        {
          additionalSetupFiles: ['@@/react-native/polyfill'],
        }
      ),
      transform,
    },
    host: {
      entry: '@/index.js',
      dependsOn: ['index'],
      app: true,
      transform,
    },
    ...{{{ bundles }}},
  },
});

`;

export default (api: IApi) => {
  const {
    utils: { winPath, Mustache, semver, resolve, lodash },
    paths,
  } = api;

  const format = createFormatter(api);

  function detectHaulPresetPath(): string {
    if (semver.valid(api.config?.reactNative?.version)) {
      const minor = semver.minor(api.config.reactNative.version);
      if (minor >= 60) {
        return dirname(require.resolve('@haul-bundler/preset-0.60/package.json'));
      } else {
        return dirname(require.resolve('@haul-bundler/preset-0.59/package.json'));
      }
    }
    throw new TypeError('未知 react-native 版本！');
  }

  api.onGenerateFiles(async () => {
    const webpack = require(resolve.sync('webpack', { basedir: paths.absSrcPath }));
    const Config = require(resolve.sync('webpack-chain', { basedir: paths.absSrcPath }));

    const env = api.env === 'production' ? 'production' : 'development';
    const webpackConfig = new Config();

    const alias = api.config.alias;
    if (typeof alias === 'object') {
      Object.keys(alias).forEach((key) => {
        const value = alias[key];
        if (value) {
          webpackConfig.resolve.alias.set(key, value);
        }
      });
    }

    // 接收插件值
    await api.applyPlugins({
      type: api.ApplyPluginsType.modify,
      key: 'chainWebpack',
      initialValue: webpackConfig,
      args: {
        webpack,
        createCSSRule: lodash.noop,
      },
    });

    // 接收用户值
    if (typeof api.config.chainWebpack === 'function') {
      await api.config.chainWebpack(webpackConfig, { env, webpack, createCSSRule: lodash.noop });
    }

    /**
     * 防止加载 umi 包 Common JS 格式的代码
     *  umi for WEB 使用 webpack treeShaking，运行时加载的是 ES Module 格式的代码：umi/dist/index.esm.js；
     *  但 RN 的模块加载方式不支持 treeShaking，会加载 umi Common JS 代码：umi/index.js，其中包含了大量 Node 工具类库。
     *  在 haul 构建时甚至会导致 out of memory。
     * 另外，使用其他方式：api.chainWebpack 或者 api.addProjectFirstLibraries "umi" alias 都会被覆盖，所以放到这里最终写 haul.config.js 时强行设置
     */
    webpackConfig.resolve.alias.set('umi', resolve.sync('umi/dist/index.esm.js', { basedir: process.env.UMI_DIR }));

    const config = webpackConfig.toConfig();
    const routes = await api.getRoutes();

    api.writeTmpFile({
      path: 'react-native/haul.config.js',
      content: format(
        Mustache.render(CONTENT, {
          haulPresetPath: winPath(detectHaulPresetPath()),
          webpackConfig: JSON.stringify(config, null, 2),
          dependencies: JSON.stringify(
            lodash.keys(
              lodash.omit(config.resolve.alias, [
                '@',
                '@@',
                'react-dom',
                'react-router-dom',
                'regenerator-runtime',
                './core/polyfill',
                './core/routes',
                '@@/core/routes',
              ]),
            ),
          ),
          bundles: bundlesToJSON(api, transformRoutesToBundle(routes)),
        }),
      ),
    });
  });
};
