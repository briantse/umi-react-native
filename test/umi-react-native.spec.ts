import { Service } from 'umi';
import rimraf from 'rimraf';
import { join } from 'path';
import { render } from '@testing-library/react-native';

const cwd = join(__dirname, 'fixtures');
const absTmp = join(cwd, '.umi-test');
const service = new Service({
  cwd,
  plugins: [require.resolve('../packages/umi-plugin-antd-react-native/src')],
  presets: [
    require.resolve('../packages/umi-preset-react-native/src'),
    // require.resolve('../packages/umi-preset-react-navigation/src'),
  ],
});

beforeAll(async () => {
  rimraf.sync(absTmp);

  await service.run({
    name: 'g',
    args: {
      _: ['g', 'tmp'],
    },
  });
});

describe('umi-plugin-antd-react-native', () => {
  it('should add runtime plugin', async () => {
    const { rootContainer } = require(join(absTmp, 'antd-react-native', 'runtime.ts'));
    const RootComponent = rootContainer(null);
    const { container } = render(RootComponent);
    expect(container.props.children).toMatchSnapshot();
  });
  it('should add babel-plugin-import options for @ant-design/react-native', async () => {
    const { plugins } = require(join(absTmp, 'babel.config.js'));
    expect(plugins).toContainEqual([
      require.resolve('babel-plugin-import'),
      {
        libraryName: '@ant-design/react-native',
      },
      '@ant-design/react-native',
    ]);
  });
});

describe('umi-preset-react-native', () => {
  it('should create babel.config.js', () => {
    const { presets, plugins } = require(join(absTmp, 'babel.config.js'));
    expect(presets).toContainEqual(require.resolve('@haul-bundler/babel-preset-react-native'));
    expect(presets).toContainEqual('babel-preset-fake');
    expect(presets).toContainEqual('babel-preset-extra-fake');
    expect(plugins).toContainEqual('babel-plugin-fake');
    expect(plugins).toContainEqual('babel-plugin-extra-fake');
  });
  it('should exports BackButton and AndroidBackButton', () => {
    const { BackButton, AndroidBackButton } = require(join(absTmp, 'react-native', 'exports.ts'));
    expect(BackButton).toBeDefined();
    expect(AndroidBackButton).toBeDefined();
  });
});
