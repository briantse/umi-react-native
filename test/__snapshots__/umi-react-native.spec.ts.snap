// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`umi-plugin-antd-react-native should write antd-react-native/runtime.ts 1`] = `
<Provider
  theme={
    Object {
      "color_text_base": "#333333",
      "fill_base": "#f6f6f6",
    }
  }
/>
`;
